streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java

```diff
@@ -58,25 +58,24 @@
     private boolean transactionInitialized = false;
 
     public StreamsProducer(final Producer<byte[], byte[]> producer,
-                           final boolean eosEnabled,
-                           final LogContext logContext,
-                           final String applicationId) {
-        log = logContext.logger(getClass());
+                           final String applicationId,
+                           final LogContext logContext) {
+        log = Objects.requireNonNull(logContext, "logContext cannot be null").logger(getClass());
         logPrefix = logContext.logPrefix().trim();
 
         this.producer = Objects.requireNonNull(producer, "producer cannot be null");
         this.applicationId = applicationId;
-        this.eosEnabled = eosEnabled;
+        this.eosEnabled = applicationId != null;
```

Comment:
Simplification to avoid passing in `eosEnabled` and reducing constructor parameter list -- we just piggy back on the `application.id` that shall be `null` for non-eos.

Reply:
It seems a bit roundabout to have to remember we should send a `null` `application.id` as the constructor argument to indicate that eos is enabled. What's wrong with saying "eos is enabled" when you want eos to be enabled?

Reply:
It a personal preference I guess. But seems you don't like it. Will revert it.